File: examples/gom.py
#!/usr/bin/python3


from gi.types import GObjectMeta
from gi.repository import GLib
from gi.repository import GObject
from gi.repository import Gom


# Need a metaclass until we get something like _gclass_init_
#     https://bugzilla.gnome.org/show_bug.cgi?id=701843
class ItemResourceMeta(GObjectMeta):
    def __init__(cls, name, bases, dct):
        super(ItemResourceMeta, cls).__init__(name, bases, dct)
        cls.set_table("items")
        cls.set_primary_key("id")
        cls.set_notnull("name")


class ItemResource(Gom.Resource, metaclass=ItemResourceMeta):
    id = GObject.Property(type=int)
    name = GObject.Property(type=str)


if __name__ == '__main__':
    # Connect to the database
    adapter = Gom.Adapter()
    adapter.open_sync(":memory:")

    # Create the table
    repository = Gom.Repository(adapter=adapter)
    repository.automatic_migrate_sync(1, [ItemResource])

    # Insert an item
    item = ItemResource(repository=repository, name="item1")
    item.save_sync()

    # Fetch the item back
    item = repository.find_one_sync(ItemResource, None)
    assert item.id == 1
    assert item.name == 'item1'

    # Insert a new item
    item = ItemResource(repository=repository, name="item2")
    item.save_sync()

    # Fetch them all with a None filter, ordered by name
    names = ['item2', 'item1']
    sorting = Gom.Sorting(ItemResource, "name", Gom.SortingMode.DESCENDING)
    group = repository.find_sorted_sync(ItemResource, None, sorting)
    count = len(group)
    assert count == 2

    group.fetch_sync(0, count)
    for i, item in enumerate(group):
        assert item.name == names[i]

    # Fetch only one of them with a filter, asynchronously
    loop = GLib.MainLoop()

    def fetch_cb(group, result, user_data):
        group.fetch_finish(result)

        item = group[0]
        assert item.name == "item2"

        # Close the database
        adapter.close_sync()

        loop.quit()

    def find_cb(repository, result, user_data):
        group = repository.find_finish(result)

        count = len(group)
        assert count == 1

        group.fetch_async(0, count, fetch_cb, None)

    filter = Gom.Filter.new_eq(ItemResource, "name", "item2")
    group = repository.find_async(ItemResource, filter, find_cb, None)

    loop.run()

File: examples/gom.js
#!/usr/bin/gjs

const Lang = imports.lang;
const GLib = imports.gi.GLib;
const GObject = imports.gi.GObject;
const GIRepository = imports.gi.GIRepository;
const Gom = imports.gi.Gom;
const System = imports.system;

const INT32_MAX = (2147483647);

const ItemClass = new Lang.Class({
    Name: 'Item',
    Extends: Gom.Resource,

    Properties: {
        'id':  GObject.ParamSpec.int('id', 'ID',
                                     'An ID', GObject.ParamFlags.READABLE | GObject.ParamFlags.WRITABLE | GObject.ParamFlags.CONSTRUCT,
                                     0, INT32_MAX, 0),
        'url': GObject.ParamSpec.string('url', 'URL',
					'A URL',
					GObject.ParamFlags.READABLE | GObject.ParamFlags.WRITABLE | GObject.ParamFlags.CONSTRUCT,
					''),
    },

    _init: function(params) {
        this.parent(params);

        Gom.Resource.set_table.call(this, 'items');
        Gom.Resource.set_primary_key.call(this, 'id');
    },

    _instance_init: function() {
    },
});

// Open
let adapter = new Gom.Adapter;
adapter.open_sync('/tmp/gom-js-test.db');
let repository = new Gom.Repository({adapter: adapter});

let item = new ItemClass({ 'repository': repository });
//item.url = 'http://www.gnome.org';
item.url = 'http://www.gnome.org';
print (item.id);
item.id = 0;
print (item.url);

// Migrate
let object_types = [ ItemClass ];
repository.automatic_migrate_sync(2, object_types);

// Add item
let ret = item.save_sync();
print("New item ID:", item.id, "URL:", item.url);

// Close
adapter.close_sync();

// Open
let adapter = new Gom.Adapter;
adapter.open_sync('/tmp/gom-js-test.db');
let repository = new Gom.Repository({adapter: adapter});

// Find item
let filter = Gom.Filter.new_eq(ItemClass, "id", 1);
let found_item = repository.find_one_sync(ItemClass, filter);

print (found_item.url);

// Close
adapter.close_sync();
